src/services/ftp/backend.rs

```diff
@@ -0,0 +1,440 @@
+            enable_secure,
+        })
+    }
+}
+
+/// Backend is used to serve `Accessor` support for ftp.
+#[derive(Clone)]
+pub struct Backend {
+    endpoint: String,
+    root: String,
+    client: Arc<Mutex<FtpStream>>,
+    enable_secure: bool,
+}
+
+impl Debug for Backend {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Backend")
+            .field("endpoint", &self.endpoint)
+            .field("root", &self.root)
+            .field("tls", &self.enable_secure)
+            .finish()
+    }
+}
+
+impl Backend {
+    pub(crate) fn from_iter(it: impl Iterator<Item = (String, String)>) -> Result<Self> {
+        let mut builder = Builder::default();
+
+        for (k, v) in it {
+            let v = v.as_str();
+            match k.as_ref() {
+                "root" => builder.root(v),
+                _ => continue,
+            };
+        }
+        builder.build()
+    }
+}
+
+#[async_trait]
+impl Accessor for Backend {
+    async fn create(&self, args: &OpCreate) -> Result<()> {
+        let path = args.path();
+        if args.mode() == ObjectMode::FILE {
+            //let mut ftp_stream = self.ftp_connect()?;
+            self.client
+                .try_lock()
+                .unwrap()
+                .put_file(&path, &mut "".as_bytes())
+                .map_err(|e| {
+                    other(ObjectError::new(
+                        Operation::Create,
+                        path,
+                        anyhow!("put request: {e:?}"),
+                    ))
+                })?;
+            return Ok(());
+        }
+
+        if args.mode() == ObjectMode::DIR {
+            self.client.try_lock().unwrap().mkdir(&path).map_err(|e| {
+                other(ObjectError::new(
+                    Operation::Create,
+                    path,
+                    anyhow!("mkdir request: {e:?}"),
+                ))
+            })?;
+
+            return Ok(());
+        }
+
+        unreachable!()
+    }
+
+    async fn read(&self, args: &OpRead) -> Result<BytesReader> {
+        let path = args.path();
+        let mut guard = self.client.lock().await;
+
+        let mut stream = guard.retr_as_stream(path).map_err(|e| {
+            other(ObjectError::new(
+                Operation::Read,
+                path,
+                anyhow!("retrieve request: {e:?}"),
+            ))
+        })?;
+
+        if let Some(offset) = args.offset() {
+            copy(&mut stream.by_ref().take(offset), &mut sink())?;
+        }
+
+        let mut buf = Vec::new();
+        let r = match args.size() {
+            None => {
+                stream.by_ref().read_to_end(&mut buf)?;
+                Box::new(Cursor::new(buf))
+            }
+            Some(size) => {
+                stream.by_ref().take(size).read_to_end(&mut buf)?;
+                Box::new(Cursor::new(buf))
+            }
+        };
+
+        guard.finalize_retr_stream(stream).map_err(|e| {
```

Comment:
So We can't return the stream to users directly except we start a new FtpStream for every read operation? Seems our previous goes wrong: reusing the same FTP connection is not a good choice.

Can you raise an issue for this problem so we can address it in the future?

Reply:
I think it is still worth to use the same connection.

FTP session operates using two channels between the FTP client and server: a `command channel` and a `data channel`. Command channel is used for transferring command to server. Data channel is used for data transferring.

The FtpStream we persist in `backend` is the **`command channel`**. The stream that we finalized above is the data stream(or `data channel`). For every new command,  a new data channel will be established between client and sever, this is unavoidable due to the property of FTP protocol.

At least we don't need to reestablish command channel for every operation.